### PR TITLE
docs: Document design for transaction signer event loop

### LINKED ISSUE #221: [Design]: Signer event loop
# Design - Signer event loop

### Context
While we have a high level design for the signer in #44, along with a detailed state model in #60, there is a lot of details that goes into the actual signer event loop and how we intend the signer to react to state updates and messages from other signers. We should write down some flow-charts illustrating how we intend the signer to behave before beginning the implementation of it.

### Definition of done
We have a detailed design with appropriate illustrations showcasing from a high-level view how the signer event loop is intended to work.

### PR DESCRIPTION
**Pull request by netrome**
closes #221 

The documentation can be rendered locally using `cargo doc --open`. The mermaid charts should be rendered in the docs thanks to [aquamarine](https://github.com/mersinvald/aquamarine).

### CODE REVIEW

File: signer/src/transaction_signer.rs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+///     PD --> BD(Broadcast decision)
+///     BD --> NR
+/// ```
+///
+/// ## On signer message
+/// When the signer receives a signer message, it needs to do a few different things
+/// depending on the type of the message.
+///
+/// - **Signer decision**: When receiving a signer decision, the transaction signer
+/// only need to persist the decision to the signer storage.
```
Comment (djordon) [inline signer/src/transaction_signer.rs:36]:
```suggestion
/// only needs to persist the decision to its database.
```

File: signer/src/transaction_signer.rs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+/// - **Stacks sign request**: When receiving a request to sign a stacks transaction,
+/// the signer must verify that it has decided to sign the transaction, and if positive,
```
Comment (djordon) [inline signer/src/transaction_signer.rs:38]:
```suggestion
/// the signer must verify that it has decided to sign the transaction, and if it has,
```

File: signer/src/transaction_signer.rs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+/// send a transaction signature back over the network.
+/// - **Bitcoin sign request**: When receiving a request to sign a bitcoin transaction,
+/// the signer must verify that it has decided to accept all request that the
```
Comment (djordon) [inline signer/src/transaction_signer.rs:41]:
```suggestion
/// the signer must verify that it has decided to accept all requests that the
```

File: signer/src/transaction_signer.rs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+/// transaction fulfills. Once verified, the transaction signer creates a dedicated
```
Comment (djordon) [inline signer/src/transaction_signer.rs:42]: Is the transaction signer always the coordinator?

File: signer/src/transaction_signer.rs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+/// WSTS state machine to participate in a signing round for this transaction. Thereafter,
+/// the signer sends a bitcoin transaction sign ack message back over the network to signal
+/// its readiness.
+/// - **WSTS message**: When receiving a WSTS message, the signer will look up the corresponding
+/// state machine and relay the WSTS message to it.
```
Comment (djordon) [inline signer/src/transaction_signer.rs:47]:
```suggestion
/// - **WSTS message**: When receiving a WSTS message, the signer will look up the corresponding
/// internal state machine process and relay the WSTS message to it.
```

**Review by djordon — COMMENTED** [review]
Nice, I didn't know about aquamarine, very nice touch.

My substantive comments are about detailing more of the specifics: what constitutes a storage update; who is the transaction signer? I'm not attached to the suggested modifications, I just figured I'd give an example of what I was thinking.

File: signer/src/transaction_signer.rs [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+#[cfg_attr(doc, aquamarine::aquamarine)]
+/// # Transaction signer event loop
+///
+/// This struct contains the implemetnation of the transaction signer logic.
+/// The event loop subscribes to storage update notifications from the block observer,
+/// and listens to signer messages from the signer entwork.
+///
+/// ## On storage update
+/// When the signer receives a storage update notification, it must go over each of the pending
+/// requests and decide wheter to accept or reject it. The decision is then persisted
```
[the remark this answers is not in the document]
Comment (netrome) [reply signer/src/transaction_signer.rs:17]: Thanks for spotting!

File: signer/src/transaction_signer.rs [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+#[cfg_attr(doc, aquamarine::aquamarine)]
+/// # Transaction signer event loop
+///
+/// This struct contains the implemetnation of the transaction signer logic.
+/// The event loop subscribes to storage update notifications from the block observer,
+/// and listens to signer messages from the signer entwork.
+///
+/// ## On storage update
+/// When the signer receives a storage update notification, it must go over each of the pending
```
[the remark this answers is not in the document]
Comment (netrome) [reply signer/src/transaction_signer.rs:16]: Right, I can clarify this. This is the notification from the block observer that the block observer has inserted more block data - which implies that at least one new bitcoin block has been created. I'll elaborate on this in the docs.

File: signer/src/transaction_signer.rs [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+///     FPR --> NR(Next request)
+///     NR --> |deposit/withdraw| DAR(Decide to accept/reject)
+///     NR ----> |none| DONE{Done}
+///     DAR --> PD(Persist decision)
+///     PD --> BD(Broadcast decision)
+///     BD --> NR
+/// ```
+///
+/// ## On signer message
+/// When the signer receives a signer message, it needs to do a few different things
```
[the remark this answers is not in the document]
Comment (netrome) [reply signer/src/transaction_signer.rs:32]: Sure thing.

File: signer/src/transaction_signer.rs (hunk 4 above)
Comment (netrome) [reply signer/src/transaction_signer.rs:43]: I view the coordinator is a separate entity with it's own event loop (as hinted in #44), but I am open to other designs. I created #222 to write down a design for the coordinator event loop.

File: signer/src/transaction_signer.rs (hunk 5 above)
Comment (netrome) [reply signer/src/transaction_signer.rs:48]: I don't think it's a separate process, so perhaps "relay" is the wrong word here.

File: signer/src/transaction_signer.rs (hunk 4 above)
Comment (djordon) [reply signer/src/transaction_signer.rs:43]: We spoke offline about this and the transaction signer process/code path is supposed to be different from the coordinator. That design makes sense to me. Resolved!

File: signer/src/transaction_signer.rs (hunk 5 above)
Comment (AshtonStephens) [reply signer/src/transaction_signer.rs:48]: `dispatch`?

**Review by AshtonStephens — APPROVED** [review]
LGTM after @djordon's suggestions.

File: signer/src/transaction_signer.rs (hunk 5 above)
Comment (netrome) [reply signer/src/transaction_signer.rs:48]: Sounds good, thanks for suggesting :)

**Comment by netrome** [comment]
> Nice, I didn't know about aquamarine, very nice touch.
> 
> My substantive comments are about detailing more of the specifics: what constitutes a storage update; who is the transaction signer? I'm not attached to the suggested modifications, I just figured I'd give an example of what I was thinking.

Me neither until today :sweat_smile: I just figured there had to be a solution to render mermaid charts inline in rustdoc.

**Comment by netrome** [comment]
Thanks for the feedback! :pray: Merging this now.

**Review by djordon — APPROVED** [review]
Looks good!

**Pull request merged**